modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1245 @@
+          and if it is empty, we attempt to become the work-stealer; if someone else is the
+          current work stealer we assist them instead and lift an element for ourselves.
+
+          Furthermore, in this phase we loop indefinitely until we are 100% certain it is
+          empty or we get an item, so introduce some backoff here.
+        */
+        when REMOVE_WORST_CASE {
+          while true {
+            ref segment = segments[idx];
+
+            select segment.currentStatus {
+              // Quick acquire
+              when STATUS_UNLOCKED {
+                if segment.acquireWithStatus(STATUS_REMOVE) {
+                  // We're lucky; another element has been added to the current segment,
+                  // take it and leave like normal...
+                  if !segment.isEmpty {
+                    var (hasElem, elem) : (bool, eltType) = segment.takeElement();
+                    segment.releaseStatus();
+                    return (hasElem, elem);
+                  }
+
+                  if parentHandle.targetLocales.size == 1 {
+                    segment.releaseStatus();
+                    return (false, _defaultOf(eltType));
+                  }
+
+                  // Attempt to become the sole work stealer for this node. If we
+                  // do not, we spin until they finish. We need to release the lock
+                  // on our segment so our segment may be load balanced as well.
+                  if loadBalanceInProgress.testAndSet() {
+                    segment.releaseStatus();
+
+                    loadBalanceInProgress.waitFor(false);
+                    var notEmpty = loadBalanceResult.read();
+                    if !notEmpty {
+                      return (false, _defaultOf(eltType));
+                    }
+
+                    // Reset our phase and scan for more elements...
+                    phase = REMOVE_BEST_CASE;
+                    break;
+                  }
+
+                  // We are the sole work stealer, and so it is our responsibility
+                  // to balance the load for our node. We fork-join new worker
+                  // tasks that will check horizontally across each node (as in
+                  // across each segment with the same index), and vertically across
+                  // each segment (each segment in a node). Horizontally, we steal
+                  // at most a % of work from other nodes to give to ourselves.
+                  // As load balancer, we also are the only one who knows whether
+                  // or not all bags are empty.
+                  var isEmpty : atomic bool;
+                  isEmpty.write(true);
+                  segment.releaseStatus();
+                  coforall segmentIdx in 0..#here.maxTaskPar {
+                    var stolenWork : [{0..#numLocales}] (int, c_ptr(eltType));
+                    coforall loc in parentHandle.targetLocales {
+                      if loc != here then on loc {

```

Comment:
The reason why we skip `if loc != here` is because we do not want to steal from ourselves.